subprojects/core/src/main/java/org/gradle/execution/plan/DefaultExecutionPlan.java

```diff
@@ -625,12 +630,22 @@ public Node selectNext(WorkerLeaseRegistry.WorkerLease workerLease, ResourceLock
                 foundReadyNode = true;
                 MutationInfo mutations = getResolvedMutationInfo(node);
 
+                if (!tryLockProjectFor(node)) {
+                    resourceLockState.releaseLocks();
+                    LOGGER.debug("Cannot acquire project lock for node {}.", node);
+                    continue;
+                } else if (!tryLockSharedResourceFor(node)) {
+                    resourceLockState.releaseLocks();
+                    LOGGER.debug("Cannot acquire shared resource lock for node {}.", node);
+                    continue;
+                } else if (!workerLease.tryLock()) {
+                    resourceLockState.releaseLocks();
+                    LOGGER.debug("Cannot acquire worker lease lock for node {}.", node);
+                    continue;
                 // TODO: convert output file checks to a resource lock
-                if (!tryLockProjectFor(node)
-                    || !tryLockSharedResourceFor(node)
-                    || !workerLease.tryLock()
-                    || !canRunWithCurrentlyExecutedNodes(node, mutations)) {
+                } else if (!canRunWithCurrentlyExecutedNodes(node, mutations)) {
                     resourceLockState.releaseLocks();
+                    LOGGER.debug("Node {} cannot run with currently running nodes {}.", node, runningNodes);
                     continue;
                 }
```

Comment:
Can we avoid having to repeat `resourceLockState.releaseLocks()`? It seems too easy to forget to do it should a new case be added.

Reply:
Done.